x/ibchooks/marker_hooks.go

```diff
@@ -0,0 +1,226 @@
+package ibchooks
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+
+	sdkerrors "cosmossdk.io/errors"
+
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+	ibckeeper "github.com/cosmos/ibc-go/v6/modules/core/keeper"
+	tendermintclient "github.com/cosmos/ibc-go/v6/modules/light-clients/07-tendermint/types"
+
+	"github.com/provenance-io/provenance/x/ibchooks/types"
+	markerkeeper "github.com/provenance-io/provenance/x/marker/keeper"
+	markertypes "github.com/provenance-io/provenance/x/marker/types"
+)
+
+type MarkerHooks struct {
+	MarkerKeeper *markerkeeper.Keeper
+}
+
+func NewMarkerHooks(markerkeeper *markerkeeper.Keeper) MarkerHooks {
+	return MarkerHooks{
+		MarkerKeeper: markerkeeper,
+	}
+}
+
+// ProperlyConfigured returns false when marker hooks are configured incorrectly
+func (h MarkerHooks) ProperlyConfigured() bool {
+	return h.MarkerKeeper != nil
+}
+
+// AddUpdateMarker will add or update ibc Marker with transfer authorities
+func (h MarkerHooks) AddUpdateMarker(ctx sdktypes.Context, packet exported.PacketI, ibcKeeper *ibckeeper.Keeper) error {
+	var data transfertypes.FungibleTokenPacketData
+	if err := json.Unmarshal(packet.GetData(), &data); err != nil {
+		return err
+	}
+	ibcDenom := MustExtractDenomFromPacketOnRecv(packet)
+	if strings.HasPrefix(ibcDenom, "ibc/") {
+		markerAddress, err := markertypes.MarkerAddress(ibcDenom)
+		if err != nil {
+			return err
+		}
+		marker, err := h.MarkerKeeper.GetMarker(ctx, markerAddress)
+		if err != nil {
+			return err
+		}
+		var transferAuthAddrs []sdktypes.AccAddress
+		transferAuthAddrs, coinType, err := ProcessMarkerMemo(data.GetMemo())
+		if err != nil {
+			return err
+		}
+		if marker == nil {
```

Comment:
Same comment as above. You can remove this indentation by making this the happy path. It's not super important, but will help improve clarity.

```
if marker != nil {
  if err = ResetMarkerAccessGrants(transferAuthAddrs, marker); err != nil {
    return err
  }
h.MarkerKeeper.SetMarker(ctx, marker)
return nil
}
```

Reply:
True, I also made a few more refactors to the method.  Extracted the rest of the logic into it's own method.  This makes it clearer what is happening after `marker != nil`, because the method name says what it does along with method comments instead adding comments in the middle of method. 
It seemed a little confusing after moving things around and the method was getting a bit long.  I hope this makes it easier to understand.